

<h1 font size = 5>Lab 4 Từ Hiểu biết đến Chuẩn bị</font></h1>

Thời lượng: 120 phút

## Giới thiệu

Trong lab này, chúng ta sẽ tiếp tục tìm hiểu về phương pháp luận khoa học dữ liệu và tập trung vào các giai đoạn **Hiểu biết dữ liệu** và **Chuẩn bị dữ liệu**.

## Mục lục

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Tóm tắt lại](#0) <br>
2. [Hiểu biết dữ liệu](#2) <br>
3. [Chuẩn bị dữ liệu](#4) <br>
</div>
<hr>

# Tóm tắt lại <a id="0"></a>

Trong bài Lab **Từ Yêu cầu đến Thu thập**, chúng ta biết rằng dữ liệu chúng ta cần để trả lời câu hỏi được phát triển trong giai đoạn Hiểu biết nghiệp vụ (cụ thể là *chúng ta có thể tự động hóa quy trình xác định nguồn gốc ẩm thực của một công thức nhất định không?*) đã sẵn sàng. Nhà nghiên cứu Yong-Yeol Ahn đã tìm kiếm hàng chục nghìn công thức nấu ăn (các món ăn và nguyên liệu) từ ba trang web khác nhau, đó là:

<img src = "https://ibm.box.com/shared/static/4fruwan7wmjov3gywiz3swlojw0srv54.png" width=500>

www.allrecipes.com

<img src = "https://ibm.box.com/shared/static/cebfdbr22fjxa47lltp0bs533r103g0z.png" width=500>

www.epicurious.com

<img src = "https://ibm.box.com/shared/static/epk727njg7xrz49pbkpkzd05cm5ywqmu.png" width=500>

www.menupan.com

Để biết thêm thông tin về Yong-Yeol Ahn và nghiên cứu của anh ấy, bạn có thể đọc bài báo của anh ấy trên [Flavor Network and the Principles of Food Pairing](http://yongyeol.com/papers/ahn-flavornet-2011.pdf).

Chúng tôi cũng đã thu thập dữ liệu và đặt nó trên máy chủ IBM để thuận tiện cho bạn.

------------

# Tìm hiểu dữ liệu <a id="2"></a>

<img src="https://ibm.box.com/shared/static/89geb3m0ge1z73s92hl8o8wdcpcrggtz.png" width=500>

<strong> Lưu ý quan trọng:</strong> Xin lưu ý rằng bạn không cần biết cách lập trình trong Python. Đoạn code sau đây nhằm minh họa các giai đoạn hiểu dữ liệu và chuẩn bị dữ liệu, vì vậy sẽ không có vấn đề gì nếu bạn chưa hiểu từng dòng code riêng lẻ. Chúng ta có một khóa học đầy đủ về lập trình bằng Python, <a href="http://cocl.us/PY0101EN_DS0103EN_LAB3_PYTHON_Coursera"> <strong> Python cho Khoa học dữ liệu </strong> </a>, cũng được cung cấp trên Coursera. Vì vậy, hãy hoàn thành khóa học Python nếu bạn muốn học cách lập trình bằng Python.

### Sử dụng notebook này:

Để chạy bất kỳ cell code nào sau đây, bạn có thể nhập **Shift + Enter** để tính toán code  trong một cell.

Cài đặt phiên bản Python.

In [1]:
# check Python version
!python -V

Python 3.10.12


Tải xuống thư viện và các tệp đi kèm mà chúng ta sẽ cần để chạy lab này.

In [2]:
import pandas as pd # import library to read data into dataframe
pd.set_option('display.max_columns', None)
import numpy as np # import numpy library
import re # import library for regular expression

Tải xuống dữ liệu từ máy chủ IBM và đọc nó vào khung dữ liệu (dataframe) *pandas*.

In [3]:
# Điền code vào đây
recipes = pd.read_csv("https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/data/recipes.csv")

print("Data read into dataframe!") # takes about 30 seconds

Data read into dataframe!


Nhấp đúp vào __đây__ để xem lời giải.
<!--
pd.read_csv
-->

Hiển thị một vài hàng đầu tiên.

In [4]:
recipes[:5]

country almond angelica anise anise_seed apple apple_brandy apricot  \
0  Vietnamese     No       No    No         No    No           No      No   
1  Vietnamese     No       No    No         No    No           No      No   
2  Vietnamese     No       No    No         No    No           No      No   
3  Vietnamese     No       No    No         No    No           No      No   
4  Vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0       No        No        No        No      No    No           No   No   
1       No        No        No        No      No    No           No   No   
2       No        No        No        No      No    No           No   No   
3       No        No        No        No      No    No           No   No   
4       No        No        No        No      No    No           No   No   

  banana barley bartlett_pear basil bay bean beech beef beef_broth beef_liver  \
0     No     No            No   Yes  No   No    No   No         No         No   
1     No     No            No    No  No   No    No   No         No         No   
2     No     No            No    No  No   No    No   No         No         No   
3     No     No            No   Yes  No  Yes    No   No        Yes         No   
4     No     No            No    No  No   No    No   No         No         No   

  beer beet bell_pepper bergamot berry bitter_orange black_bean black_currant  \
0   No   No          No       No    No            No         No            No   
1   No   No          No       No    No            No         No            No   
2   No   No          No       No    No            No         No            No   
3   No   No          No       No    No            No         No            No   
4   No   No          No       No    No            No         No            No   

  black_mustard_seed_oil black_pepper black_raspberry black_sesame_seed  \
0                     No           No              No                No   
1                     No          Yes              No                No   
2                     No           No              No                No   
3                     No           No              No                No   
4                     No           No              No                No   

  black_tea blackberry blackberry_brandy blue_cheese blueberry bone_oil  \
0        No         No                No          No        No       No   
1        No         No                No          No        No       No   
2        No         No                No          No        No       No   
3        No         No                No          No        No       No   
4        No         No                No          No        No       No   

  bourbon_whiskey brandy brassica bread broccoli brown_rice brussels_sprout  \
0              No     No       No    No       No         No              No   
1              No     No       No    No       No         No              No   
2              No     No       No    No       No         No              No   
3              No     No       No    No       No         No              No   
4              No     No       No    No       No         No              No   

  buckwheat butter buttermilk cabbage cabernet_sauvignon_wine cacao  \
0        No     No         No      No                      No    No   
1        No     No         No      No                      No    No   
2        No     No         No      No                      No    No   
3        No     No         No      No                      No    No   
4        No     No         No      No                      No    No   

  camembert_cheese cane_molasses caraway cardamom carnation carob carrot  \
0               No            No      No       No        No    No    Yes   
1               No            No      No       No        No    No     No   
2               No            No      No       No        No    No     No   
3   

Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes.head()
-->

Lấy chiều từ khung dữ liệu.

In [5]:
recipes.shape

(57691, 384)

Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes.shape
-->

Vì vậy, tập dữ liệu của chúng ta gồm 57.691 công thức nấu ăn. Mỗi hàng đại diện cho một công thức và với mỗi công thức, món ăn tương ứng được ghi lại cũng như xem 384 thành phần nguyên liệu có tồn tại trong công thức hay không, bắt đầu bằng hạnh nhân (almond) và kết thúc bằng bí ngòi (zuchini).

Chúng ta biết rằng công thức sushi cơ bản bao gồm các thành phần:
* cơm
* xì dầu
* wasabi (mù tạt)
* một ít cá/rau

Hãy kiểm tra xem các thành phần này có tồn tại trong khung dữ liệu của chúng ta không:

In [6]:
ingredients = list(recipes.columns.values)

print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(rice).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(wasabi).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(soy).*")).search(ingredient)] if match])

['brown_rice', 'licorice', 'rice']
['wasabi']
['soy_sauce', 'soybean', 'soybean_oil']


Có các thành phần này!

* cơm tồn tại dưới dạng rice.
* mù tạt tồn tại dưới dạng wasabi.
* đậu nành tồn tại dưới dạng soy_sauce.

Vì vậy, có thể nếu một công thức có cả ba thành phần: cơm, mù tạt và đậu nành, thì chúng ta có thể tự tin nói rằng công thức đó là một món ăn **Nhật Bản**. Hãy ghi nhớ điều này!

----------------

# Chuẩn bị Dữ liệu <a id="4"></a>

<img src="https://ibm.box.com/shared/static/lqc2j3r0ndhokh77mubohwjqybzf8dhk.png" width=500>

Trong phần này, chúng ta sẽ chuẩn bị dữ liệu cho giai đoạn tiếp theo trong phương pháp luận khoa học dữ liệu, đó là mô hình hóa. Giai đoạn này bao gồm việc khám phá dữ liệu sâu hơn và đảm bảo rằng dữ liệu ở định dạng phù hợp cho thuật toán học máy mà chúng ta đã chọn trong giai đoạn tiếp cận phân tích, đó là cây quyết định.

Đầu tiên, hãy xem dữ liệu để xem liệu nó có cần làm sạch (cleaning) hay không.

In [8]:
# Điền code vào
recipes["country"].value_counts # frequency table

<bound method IndexOpsMixin.value_counts of 0        Vietnamese
1        Vietnamese
2        Vietnamese
3        Vietnamese
4        Vietnamese
            ...    
57686         Japan
57687         Japan
57688         Japan
57689         Japan
57690         Japan
Name: country, Length: 57691, dtype: object>

Nhấp đúp vào __đây__ để xem lời giải.
<!--
.value_counts()
-->

Bằng cách nhìn vào bảng trên, chúng ta có thể đưa ra các nhận xét sau:

1. Cột Cuisine  (Món ăn) được gắn nhãn là Country (Quốc gia) là không chính xác.
2. Tên món ăn không nhất quán vì một số không viết hoa chữ cái đầu.
3. Một số món ăn bị trùng lặp dưới dạng biến tấu của tên quốc gia, chẳng hạn như Vietnam và Vietnamese.
4. Một số nền ẩm thực có rất ít công thức nấu ăn.

#### Hãy khắc phục những vấn đề này.

Sửa tên của cột hiển thị tên món ăn (cuisine)

In [9]:
column_names = recipes.columns.values

column_names[0] =  "cuisine"

recipes.columns = column_names

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      Vietnamese     No       No    No         No    No           No      No   
1      Vietnamese     No       No    No         No    No           No      No   
2      Vietnamese     No       No    No         No    No           No      No   
3      Vietnamese     No       No    No         No    No           No      No   
4      Vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686       Japan     No       No    No         No    No           No      No   
57687       Japan     No       No    No         No    No           No      No   
57688       Japan     No       No    No         No    No           No      No   
57689       Japan     No       No    No         No    No           No      No   
57690       Japan     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57686       No        No        No        No      No    No           No   No   
57687       No        No        No        No      No    No           No   No   
57688       No        No        No        No      No    No           No   No   
57689       No        No        No        No      No   Yes           No   No   
57690       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57686     No     No            No    No  No   No    No   No         No   
57687     No     No            No    No  No   No    No   No         No   
57688     No    Yes            No    No  No   No    No   No         No   
57689     No     No            No    No  No   No    No   No         No   
57690     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57686         No   No   No          No       No    No            No   
57687         No   No   No          No       No    No            No   
57688         No   No   No          No       No    No            No   
57689         No   No   No          No       No    No            No   
57690         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

Nhấp đúp vào __đây__ để xem lời giải.
<!--
column_names[0] =  "cuisine"
-->

Đảm bảo tên của các nền ẩm thực viết bằng chữ thường.

In [11]:
# Điền code vào đây
recipes["cuisine"] = recipes["cuisine"].str.lower()

Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes["cuisine"] = recipes["cuisine"].str.lower()-->

Để tên của các nền ẩm thực đồng nhất với nhau.

In [ ]:
recipes.loc[recipes["cuisine"] == "austria", "cuisine"] = "austrian"
recipes.loc[recipes["cuisine"] == "belgium", "cuisine"] = "belgian"
recipes.loc[recipes["cuisine"] == "china", "cuisine"] = "chinese"
recipes.loc[recipes["cuisine"] == "canada", "cuisine"] = "canadian"
recipes.loc[recipes["cuisine"] == "netherlands", "cuisine"] = "dutch"
recipes.loc[recipes["cuisine"] == "france", "cuisine"] = "french"
recipes.loc[recipes["cuisine"] == "germany", "cuisine"] = "german"
recipes.loc[recipes["cuisine"] == "india", "cuisine"] = "indian"
recipes.loc[recipes["cuisine"] == "indonesia", "cuisine"] = "indonesian"
recipes.loc[recipes["cuisine"] == "iran", "cuisine"] = "iranian"
recipes.loc[recipes["cuisine"] == "italy", "cuisine"] = "italian"
recipes.loc[recipes["cuisine"] == "japan", "cuisine"] = "japanese"
recipes.loc[recipes["cuisine"] == "israel", "cuisine"] = "jewish"
recipes.loc[recipes["cuisine"] == "korea", "cuisine"] = "korean"
recipes.loc[recipes["cuisine"] == "lebanon", "cuisine"] = "lebanese"
recipes.loc[recipes["cuisine"] == "malaysia", "cuisine"] = "malaysian"
recipes.loc[recipes["cuisine"] == "mexico", "cuisine"] = "mexican"
recipes.loc[recipes["cuisine"] == "pakistan", "cuisine"] = "pakistani"
recipes.loc[recipes["cuisine"] == "philippines", "cuisine"] = "philippine"
recipes.loc[recipes["cuisine"] == "scandinavia", "cuisine"] = "scandinavian"
recipes.loc[recipes["cuisine"] == "spain", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "portugal", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "switzerland", "cuisine"] = "swiss"
recipes.loc[recipes["cuisine"] == "thailand", "cuisine"] = "thai"
recipes.loc[recipes["cuisine"] == "turkey", "cuisine"] = "turkish"
recipes.loc[recipes["cuisine"] == "vietnam", "cuisine"] = "vietnamese"
recipes.loc[recipes["cuisine"] == "uk-and-ireland", "cuisine"] = "uk-and-irish"
recipes.loc[recipes["cuisine"] == "irish", "cuisine"] = "uk-and-irish"

recipes

Loại bỏ các nền ẩm thực có < 50 công thức.

In [13]:
# get list of cuisines to keep
recipes_counts = recipes["cuisine"].value_counts()
cuisines_indices = recipes_counts > 50

# Điền code vào đây
cuisines_to_keep = list(np.array(recipes_counts.index.values)[np.array(cuisines_indices)])

Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes_counts.index.values
-->

In [15]:
rows_before = recipes.shape[0] # number of rows of original dataframe
print("Number of rows of original dataframe is {}.".format(rows_before))

# Điền code vào đây
recipes['cuisine'].isin(cuisines_to_keep)

rows_after = recipes.shape[0] # number of rows of processed dataframe
print("Number of rows of processed dataframe is {}.".format(rows_after))

print("{} rows removed!".format(rows_before - rows_after))

Number of rows of original dataframe is 57691.
Number of rows of processed dataframe is 57691.
0 rows removed!


Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes['cuisine'].isin(cuisines_to_keep)
-->

Chuyển tất cả Yes thành 1 và No thành 0

In [19]:
# Điền code vào đây
recipes = recipes.replace(to_replace="Yes", value=1)
recipes = recipes.replace(to_replace="No", value=0)

#### Hãy phân tích dữ liệu thêm một chút để tìm hiểu dữ liệu tốt hơn và ghi nhận bất kỳ quan sát sơ bộ thú vị nào.

Chạy cell sau để nhận công thức có chứa **rice** *và* **soy** *và* **wasabi** *và* **seaweed**.

In [20]:
recipes.head()

cuisine  almond  angelica  anise  anise_seed  apple  apple_brandy  \
0  vietnamese       0         0      0           0      0             0   
1  vietnamese       0         0      0           0      0             0   
2  vietnamese       0         0      0           0      0             0   
3  vietnamese       0         0      0           0      0             0   
4  vietnamese       0         0      0           0      0             0   

   apricot  armagnac  artemisia  artichoke  asparagus  avocado  bacon  \
0        0         0          0          0          0        0      0   
1        0         0          0          0          0        0      0   
2        0         0          0          0          0        0      0   
3        0         0          0          0          0        0      0   
4        0         0          0          0          0        0      0   

   baked_potato  balm  banana  barley  bartlett_pear  basil  bay  bean  beech  \
0             0     0       0       0              0      1    0     0      0   
1             0     0       0       0              0      0    0     0      0   
2             0     0       0       0              0      0    0     0      0   
3             0     0       0       0              0      1    0     1      0   
4             0     0       0       0              0      0    0     0      0   

   beef  beef_broth  beef_liver  beer  beet  bell_pepper  bergamot  berry  \
0     0           0           0     0     0            0         0      0   
1     0           0           0     0     0            0         0      0   
2     0           0           0     0     0            0         0      0   
3     0           1           0     0     0            0         0      0   
4     0           0           0     0     0            0         0      0   

   bitter_orange  black_bean  black_currant  black_mustard_seed_oil  \
0              0           0              0                       0   
1              0           0              0                       0   
2              0           0              0                       0   
3              0           0              0                       0   
4              0           0              0                       0   

   black_pepper  black_raspberry  black_sesame_seed  black_tea  blackberry  \
0             0                0                  0          0           0   
1             1                0                  0          0           0   
2             0                0                  0          0           0   
3             0                0                  0          0           0   
4             0                0                  0          0           0   

   blackberry_brandy  blue_cheese  blueberry  bone_oil  bourbon_whiskey  \
0                  0            0          0         0                0   
1                  0            0          0         0                0   
2                  0            0          0         0                0   
3                  0            0          0         0                0   
4                  0            0          0         0                0   

   brandy  brassica  bread  broccoli  brown_rice  brussels_sprout  buckwheat  \
0       0         0      0         0           0                0          0   
1       0         0      0         0           0                0          0   
2       0         0      0         0           0                0          0   
3       0         0      0         0           0                0          0   
4       0         0      0         0           0                0          0   

   butter  buttermilk  cabbage  cabernet_sauvignon_wine  cacao  \
0       0           0        0                        0      0   
1       0           0        0                        0      0   
2       0           0        0                        0      0   
3       0           0        0                        0      0   
4       

In [21]:
check_recipes = recipes.loc[
    (recipes["rice"] == 1) &
    (recipes["soy_sauce"] == 1) &
    (recipes["wasabi"] == 1) &
    (recipes["seaweed"] == 1)
]

check_recipes

cuisine  almond  angelica  anise  anise_seed  apple  apple_brandy  \
11306    japanese       0         0      0           0      0             0   
11321    japanese       0         0      0           0      0             0   
11361    japanese       0         0      0           0      0             0   
12171       asian       0         0      0           0      0             0   
12385       asian       0         0      0           0      0             0   
13010       asian       0         0      0           0      0             0   
13159       asian       0         0      0           0      0             0   
13513    japanese       0         0      0           0      0             0   
13586    japanese       0         0      0           0      0             0   
13625  east_asian       0         0      0           0      0             0   
14495  east_asian       0         0      0           0      0             0   

       apricot  armagnac  artemisia  artichoke  asparagus  avocado  bacon  \
11306        0         0          0          0          0        1      0   
11321        0         0          0          0          0        0      0   
11361        0         0          0          0          0        1      0   
12171        0         0          0          0          0        0      0   
12385        0         0          0          0          0        1      0   
13010        0         0          0          0          0        1      0   
13159        0         0          0          0          0        1      0   
13513        0         0          0          0          0        0      0   
13586        0         0          0          0          0        0      0   
13625        0         0          0          0          0        0      0   
14495        0         0          0          0          0        0      0   

       baked_potato  balm  banana  barley  bartlett_pear  basil  bay  bean  \
11306             0     0       0       0              0      0    0     0   
11321             0     0       0       0              0      0    0     0   
11361             0     0       0       0              0      0    0     0   
12171             0     0       0       0              0      0    0     0   
12385             0     0       0       0              0      0    0     0   
13010             0     0       0       0              0      0    0     0   
13159             0     0       0       0              0      0    0     0   
13513             0     0       0       0              0      0    0     0   
13586             0     0       0       0              0      0    0     0   
13625             0     0       0       0              0      0    0     0   
14495             0     0       0       0              0      0    0     0   

       beech  beef  beef_broth  beef_liver  beer  beet  bell_pepper  bergamot  \
11306      0     0           0           0     0     0            0         0   
11321      0     0           0           0     0     0            0         0   
11361      0     0           0           0     0     0            0         0   
12171      0     0           0           0     0     0            0         0   
12385      0     0           0           0     0     0            0         0   
13010      0     0           0           0     0     0            0         0   
13159      0     0           0           0     0     0            0         0   
13513      0     0           0           0     0     0            0         0   
13586      0     0           0           0     0     0            0         0   
13625      0     0           0           0     0     0            0         0   
14495      0     0           0           0     0     0            0         0   

       berry  bitter_orange  black_bean  black_currant  \
11306      0              0           0              0   
11321      0              0           0              0   
11361      0              0           0             

Dựa trên kết quả của code trên, chúng ta  có thể phân loại tất cả các công thức nấu ăn có chứa **rice** *và* **soy** *và* **wasabi** *và* **seaweed** thành  công thức nấu ăn của **Japanese** không? Tại sao?

Câu trả lời của bạn:

Nhấp đúp vào __đây__ để xem lời giải.
<!-- Câu trả lời đúng là:
Không, bởi vì các công thức nấu ăn khác như công thức nấu ăn Châu Á và Đông Á cũng chứa các thành phần này.
-->

Hãy đếm các thành phần trong tất cả các công thức nấu ăn.

In [22]:
# sum each column
# Điền code vào đây
recipes.iloc[:, 1:].sum(axis=0)


almond        2329
angelica         1
anise          224
anise_seed      87
apple         2433
              ... 
wood            33
yam             86
yeast         3402
yogurt        1057
zucchini      1107
Length: 383, dtype: int64

Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes.iloc[:, 1:].sum(axis=0)
-->

In [29]:
ing = pd.Series({'Flour': 2, 'Sugar': 1, 'Eggs': 12})

# Now, using `ing` to create the Series for ingredients and counts
ingredient = pd.Series(ing.index.values, index = np.arange(len(ing)))
count = pd.Series(ing.values, index = np.arange(len(ing)))  # Changed from list(ing) to ing.values for clarity

# Create the DataFrame
ing_df = pd.DataFrame({'ingredient': ingredient, 'count': count})

print(ing_df.to_string())

  ingredient  count
0      Flour      2
1      Sugar      1
2       Eggs     12


Nhấp đúp vào __đây__ để xem lời giải.
<!--
pd.DataFrame(dict(ingredient = ingredient, count = count))
-->

> Add blockquote



Bây giờ chúng ta có khung dữ liệu và tổng số các thành phần trên tất cả các công thức nấu ăn. Hãy sắp xếp khung dữ liệu này theo thứ tự giảm dần.

In [30]:
ing_df.sort_values(["count"], ascending=False, inplace=True)
ing_df.reset_index(inplace=True, drop=True)

print(ing_df)

  ingredient  count
0       Eggs     12
1      Flour      2
2      Sugar      1


Nhấp đúp vào __đây__ để xem lời giải.
<!--
sort_values
-->

#### 3 thành phần nguyên liệu phổ biến nhất là gì?

Đáp án của bạn:
1. ...

2. ...

3. ...

Nhấp đúp vào __đây__ để xem lời giải.
<!-- Câu trả lời đúng là:
// 1. egg với 21.025 lần xuất hiện.
// 2. wheat với 20,781 lần xuất hiện.
// 3. butter với 20,719 lần xuất hiện.
-->

Tuy nhiên, lưu ý rằng có một vấn đề với bảng trên. Có ~ 40.000 công thức nấu ăn của Mỹ trong tập dữ liệu, có nghĩa là dữ liệu thiên về các nguyên liệu của Mỹ.

**Do đó**, hãy tính toán một bản tóm tắt khách quan hơn về các thành phần bằng cách xem xét các thành phần của mỗi món ăn.

#### Hãy tạo một *profile* (hồ sơ) cho từng nền ẩm thực.

Nói cách khác, chúng ta hãy thử tìm hiểu xem người Trung Quốc thường sử dụng những nguyên liệu nào và ẩm thực **Canada** như thế nào chẳng hạn.

In [32]:
cuisines = recipes.groupby("cuisine").mean()
cuisines.head()

almond  angelica     anise  anise_seed     apple  apple_brandy  \
cuisine                                                                        
african     0.156522  0.000000  0.000000    0.000000  0.034783      0.000000   
american    0.040598  0.000025  0.003014    0.000573  0.052055      0.000623   
asian       0.007544  0.000000  0.000838    0.002515  0.012573      0.000000   
austria     0.238095  0.000000  0.000000    0.000000  0.047619      0.000000   
bangladesh  0.000000  0.000000  0.000000    0.000000  0.000000      0.000000   

             apricot  armagnac  artemisia  artichoke  asparagus   avocado  \
cuisine                                                                     
african     0.069565    0.0000        0.0   0.000000   0.000000  0.008696   
american    0.011308    0.0001        0.0   0.006351   0.007522  0.007323   
asian       0.005029    0.0000        0.0   0.000000   0.019279  0.010897   
austria     0.142857    0.0000        0.0   0.000000   0.000000  0.000000   
bangladesh  0.000000    0.0000        0.0   0.000000   0.000000  0.000000   

               bacon  baked_potato      balm    banana    barley  \
cuisine                                                            
african     0.008696      0.000000  0.000000  0.017391  0.017391   
american    0.041220      0.000199  0.000075  0.021594  0.003885   
asian       0.013412      0.000000  0.000000  0.006706  0.029338   
austria     0.000000      0.000000  0.000000  0.000000  0.000000   
bangladesh  0.000000      0.000000  0.000000  0.000000  0.000000   

            bartlett_pear     basil       bay      bean  beech      beef  \
cuisine                                                                    
african          0.000000  0.017391  0.034783  0.008696    0.0  0.052174   
american         0.000423  0.058655  0.020697  0.028767    0.0  0.075118   
asian            0.000000  0.055323  0.011735  0.083822    0.0  0.103940   
austria          0.000000  0.000000  0.047619  0.000000    0.0  0.047619   
bangladesh       0.000000  0.000000  0.500000  0.000000    0.0  0.500000   

            beef_broth  beef_liver      beer      beet  bell_pepper  bergamot  \
cuisine                                                                         
african       0.008696    0.000000  0.000000  0.008696     0.313043  0.000000   
american      0.009988    0.000174  0.005430  0.003686     0.087846  0.000149   
asian         0.012573    0.000000  0.004191  0.000838     0.133277  0.000000   
austria       0.047619    0.000000  0.000000  0.000000     0.000000  0.000000   
bangladesh    0.000000    0.000000  0.000000  0.000000     0.000000  0.000000   

               berry  bitter_orange  black_bean  black_currant  \
cuisine                                                          
african     0.000000       0.000000    0.008696       0.000000   
american    0.003836       0.001245    0.005131       0.000199   
asian       0.000838       0.001676    0.017603       0.000000   
austria     0.000000       0.000000    0.000000       0.000000   
bangladesh  0.000000       0.000000    0.000000       0.000000   

            black_mustard_seed_oil  black_pepper  black_raspberry  \
cuisine                                                             
african                   0.000000      0.295652         0.000000   
american                  0.000050      0.147621         0.000199   
asian                     0.001676      0.118189         0.000000   
austria                   0.000000      0.047619         0.000000   
bangladesh                0.000000      0.250000         0.000000   

            black_sesame_seed  black_tea  blackberry  blackberry_brandy  \
cuisine                                                                   
african                   0.0   0.000000    0.000000           0.000000   
american                  0.0   0.000573    0.003462           0.000075   
asian                     0.0   0.005868    0.000838           0.000000   
austria      

Nhấp đúp vào __đây__ để xem lời giải.
<!--
groupby
-->

Như được hiển thị ở trên, chúng ta vừa tạo khung dữ liệu trong đó mỗi hàng là một món ăn và mỗi cột (ngoại trừ cột đầu tiên) là một thành phần và các giá trị hàng thể hiện tỷ lệ phần trăm của mỗi thành phần trong món ăn tương ứng.

**Ví dụ**:

* *almond* có mặt trên 15,65% tất cả các công thức của **người châu Phi**.
* *butter* có mặt trên 38,11% tất cả các công thức của **người Canada**.

Hãy in hồ sơ cho mỗi món ăn bằng cách hiển thị 4 thành phần nguyên liệu hàng đầu trong từng nền ẩm thực.

In [33]:
num_ingredients = 4 # define number of top ingredients to print

# define a function that prints the top ingredients for each cuisine
def print_top_ingredients(row):
    print(row.name.upper())
    row_sorted = row.sort_values(ascending=False)*100
    top_ingredients = list(row_sorted.index.values)[0:num_ingredients]
    row_sorted = list(row_sorted)[0:num_ingredients]

    for ind, ingredient in enumerate(top_ingredients):
        print("%s (%d%%)" % (ingredient, row_sorted[ind]), end=' ')
    print("\n")

# apply function to cuisines dataframe
create_cuisines_profiles = cuisines.apply(print_top_ingredients, axis=1)

AFRICAN
onion (53%) olive_oil (52%) garlic (49%) cumin (42%) 

AMERICAN
butter (41%) egg (40%) wheat (39%) onion (29%) 

ASIAN
soy_sauce (49%) ginger (48%) garlic (47%) rice (41%) 

AUSTRIA
egg (80%) wheat (80%) butter (71%) milk (47%) 

BANGLADESH
cayenne (100%) onion (100%) turmeric (100%) vegetable_oil (100%) 

BELGIUM
wheat (90%) butter (72%) cane_molasses (54%) egg (45%) 

CAJUN_CREOLE
onion (69%) cayenne (56%) garlic (48%) butter (36%) 

CANADA
wheat (39%) butter (38%) egg (35%) onion (34%) 

CARIBBEAN
onion (51%) garlic (50%) vegetable_oil (31%) black_pepper (31%) 

CENTRAL_SOUTHAMERICAN
garlic (56%) onion (54%) cayenne (51%) tomato (41%) 

CHINA
soy_sauce (70%) garlic (45%) scallion (43%) egg (39%) 

CHINESE
soy_sauce (67%) ginger (59%) garlic (56%) scallion (50%) 

EAST-AFRICAN
onion (63%) garlic (36%) ginger (36%) cardamom (27%) 

EAST_ASIAN
garlic (55%) soy_sauce (50%) scallion (49%) cayenne (47%) 

EASTERN-EUROPE
wheat (53%) egg (52%) butter (48%) onion (45%) 

EASTERNEUROP

Tại thời điểm này, chúng ta đã có thể hiểu rõ về dữ liệu và dữ liệu đã sẵn sàng và ở định dạng phù hợp để lập mô hình!

-----------

### Cảm ơn bạn đã hoàn thành lab này!

Nguồn: IBM